domain/storageprovisioning/service/entitylife.go

```diff
@@ -0,0 +1,136 @@
+// called once on the first invocation of the returned mapper.
+//
+// The returned mapper is not thread safe.
+func EntityLifeMapperFunc(
+	initialLife func(ctx context.Context) (map[string]life.Life, error),
+	lifeGetter EntityLifeGetter,
+) eventsource.Mapper {
+	var haveInitialLife bool
+	var knownLife map[string]life.Life
+	return func(
+		ctx context.Context, _ []changestream.ChangeEvent,
+	) ([]string, error) {
+		if !haveInitialLife {
+			l, err := initialLife(ctx)
+			if err != nil {
+				return nil, errors.Errorf("getting initial life for mapper: %w", err)
+			}
+			haveInitialLife = true
+			knownLife = l
+		}
+
+		latestLife, err := lifeGetter(ctx)
+		if err != nil {
+			return nil, errors.Errorf(
+				"getting latest storage entity life values: %w", err,
+			)
+		}
+
+		changes := []string(nil)
+		for k, v := range latestLife {
+			if l, has := knownLife[k]; !has || v != l {
+				changes = append(changes, k)
+			}
+			delete(knownLife, k)
+		}
+		changes = slices.AppendSeq(changes, maps.Keys(knownLife))
+		knownLife = latestLife
+		return changes, nil
+	}
+}
+
+// MakeEntityLifePrerequisites is a helper function for establishing the
+// prerequisites required for watching the life of storage entities associated
+// with a concern. A concern in this case is a machine being watched by a
+// provisioner.
+//
+// This helper function exists to make sure that the initial life values
+// reported by the initial query are used by the returned mapper for detecting
+// change. Keeping the initial query and the mapper in sync stops the
+// introduction of race conditions when reporting changes.
+//
+// If the initial query has not provided any initial values to the mapper then
+// the mapper starts from an empty set of known life values. The design is done
+// like this so that the two components never create a dead lock between one
+// another.
+//
+// This function returns a new initial query that can be supplied to the watcher
+// for seeding a set of initial values.
+func MakeEntityLifePrerequisites(
+	initialQuery eventsource.Query[map[string]life.Life],
+	lifeGetter EntityLifeGetter,
+) (eventsource.NamespaceQuery, eventsource.Mapper) {
+	// Make a buffered channel to capture the initial query values.
+	// Shimmed query is responsible for closing the channel.
+	initial := make(chan map[string]life.Life, 1)
+	shimmedInitialQuery := func(ctx context.Context, db database.TxnRunner) ([]string, error) {
+		defer close(initial)
+		initQueryData, err := initialQuery(ctx, db)
+		if err != nil {
+			return nil, errors.Errorf(
+				"running initial query for entity life values: %w", err,
+			)
+		}
+		select {
+		case initial <- initQueryData:
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+		return slices.Collect(maps.Keys(initQueryData)), nil
+	}
+
+	initialLifeForMapper := func(ctx context.Context) (map[string]life.Life, error) {
+		select {
+		case initialData := <-initial:
+			return initialData, nil
```

Comment:
Can we add one guard rail please:
```suggestion
		case initialData, ok := <-initial:
			if !ok {
				return nil, errors.ConstError("initial life can only be called once or initial query failed")
			}
			return initialData, nil
```

Reply:
Done & Done